### ISSUE TITLE
[question] servo as a desktop UI/UX layer?

### ISSUE THREAD

**Issue by leeoniya** [redox-os/redox#615]
Hey guys,

Sorry if this is a dumb question/idea, as I know very little about the servo's OS-level dependencies.

Is something like utilizing servo for a desktop/ui layer in Redox even feasible (similar to what gaia is for b2g [1] or Electron [1])? I imagine it would be a huge amount of work, but it would eliminate the need to re-invent or reimplement flexbox (eg for window tiling), CSS animations/transitions, shadows, etc. You could potentially offload the entire UI design to designers and quickly spawn a theming community. An OS where the UI/UX is 100% styleable with little friction would be incredibly cool.

[1] https://github.com/mozilla-b2g/gaia
[2] http://electron.atom.io/

**Comment by jackpot51** [comment]
We are unlikely to get around to this any time soon, porting servo would already be a massive operation

**Comment by Mouvedia** [comment]
dupe of #448

**Comment by ca1ek** [comment]
@Mouvedia not really, #448 is talking about porting Servo to Redox, this issue is talking about using it for UI if/when it's here.

**Comment by Mouvedia** [comment]
@ca1ek code wise it involves porting a lot of dependencies.

cf https://github.com/redox-os/redox/issues/448#issuecomment-171357177

**Comment by hon** [comment]
it sounds like another firefox os or webos?

**Comment by ca1ek** [comment]
I don't know much about them but I don't feel like that. HTML + CSS was designed to align text in 2D space (no matter how bad the memes make it seem). It may work nicely for this use case.

**Comment by genodeftest** [comment]
I don't think having a web engine run inefficient JavaScript is a useful way to build GUIs. I'd rather suggest not reinventing the wheel and reuse a cross-platform toolkit such as Gtk or Qt. Redox community is too small to develop just another complete GUI stack.

**Comment by leeoniya** [comment]
> I don't think having a web engine run inefficient JavaScript is a useful way to build GUIs.

Servo is composed of multiple components. The JS engine is simply one component and nothing in servo relies on it. I'm suggesting using Webrender (which is mostly glsl shaders & rust) and the parallel layout/styling engine wgich is Rust (not js). I believe GTK also uses CSS syntax in several places, btw. Javascript is not needed for any of this, however modern JS engines are more than fast enough if one were to opt for it as a scripting layer.

**Comment by ayoshi** [comment]
Theoretically It would be the best way forward, yes. Web had improved so much these days. Building another bad option like Qt or GTK, which breed monstrosities will make for another bad experience, akin to Linux on the Desktop, heck even Windows. Electron and Atom/Slack had proved to us that building interfaces for user applications should be left to those who properly can and really care about small details, and unfortunately most of those folks are writing html/css/js, people developing apps for MacOS are a notable exception. It would also allow to port amazing things with a minimum hassle (again Atom, Slack, JS terminals, who knows what else).

**Comment by Mouvedia** [comment]
@ayoshi you should be interested in the elementary OS [human interface guidelines](https://elementary.io/docs/human-interface-guidelines).

**Comment by ayoshi** [comment]
@Mouvedia yes, I've seen them, and they are a step forward in many ways. I've been fighting Linux desktop since well, twm or earlier (Redhat 4, and I don't mean RHEL). The reality is, the guidelines are important but not sufficient. The entry barrier into GTK/Qt is too high for people who would otherwise produce great interfaces, and those who can design them would simply turn away. The same would probably be correct for any other platform specific GUI library in any language which is not JS. Also, laying out and styling stuff in GUI is really, really hard if you want to do it right. There is a reason why CSS/HTML is so bloated beyond simply historical development and being initially unfit for interfaces - doing layout and presentation properly is a lot of work and requires hundreds of knobs to turn, sometimes on a level of a single pixel. Recreating the same capabilities in a library up to a required extent would arguably give us back CSS/HTML in a different disguise. That said, it's not that it can't be improved, just that the current state of the art required many years of hard thinking to mature, and a whole generation of developers is capable of using it as a tool. But I digress, this is probably off-topic.

**Comment by ritschwumm** [comment]
@ayoshi i don't think you need CSS in a different disguise. you need a constraint solver, and a way to feed rules to it.
CSS is neither an especially clean nor an especially elegant way to do this. on the contrary: even for  halfway complex guis i found it easier to write raw constraints for something like cassowary than figuring out how to make CSS do what i want.
once you have a solid foundation (cassowary might not be enough..) you can leave support for things like "flexbox" to the users of your api - it's just half a screenful of code anyway.

**Comment by rofrol** [comment]
I don't think Redox GUI needs to be rendered with Servo. It should be minimal (and OS gui should be minimal - window decorations? taskbar? what else is neede?) and later apps can be built using Servo/Webassembly/Houdini and when OS exposes needed API they will native-like. This way Redox progress will be faster in essential os development.

So file manager, music player etc. can be built on top of Redox using web technologies.

**Comment by erlend-sh** [comment]
This has also been discussed on the forum:

https://discourse.redox-os.org/t/using-servo-as-an-application-design-tool-when-it-becomes-embeddable/548?u=erlend_sh

Since this is a very open-ended debate I suggest closing this issue so discussion can continue on the forum.

**Comment by bugQ** [comment]
"another bad option like Qt or Gtk" :\
"The entry barrier into GTK/Qt is too high for people" :(

Qt remains way more portable than anything written for the web.  If the barrier for entry is too high, then high-level idiomatic bindings are the answer.  Isn't this supposed to be an OS by and for Rust users ?

**Comment by rofrol** [comment]
This one is interesting https://github.com/Thinkofname/stylish
